### DISCUSSION TITLE
Mysql database selection (db-per-tenant application)

### DISCUSSION THREAD

**Discussion by mgraf-dozuki** [prisma/prisma#24456] [Q&A]
### Question

I am working on a new application using Prisma.  We have a database-per-tenant architecture for data segregation and portability, and are using Mysql.  Number of tenants in system is ~1000 with maybe 200 using the application at any one time.

In order to create a PrismaClient for the relevant tenant of a request, I started with maintaining a pool of pools approach, caching PrismaClients for tenants as requests come in, and reusing them for later requests.  I added cleanup for unused clients after some expiration to help maintain size some.  However the number of actual db connections became too high.

Then I moved on to attempting to use a single PrismaClient, but using a client extension to try to select the tenant database in a manner similar to the RLS example (https://github.com/prisma/prisma-client-extensions/blob/main/row-level-security/script.ts).  Instead of doing the set_config() for RLS, I did a `prisma.$executeRawUnsafe('USE tenantA;')` to try and select the database for the tenant.  However in this case, it appears that `$executeRawUnsafe()` will send this as a prepared statement to Mysql (presumably because it is in a transaction using prepared statements) and `USE` is not supported in prepared statements.  Thus I get the following error:

`Raw query failed. Code: '1295'. Message: 'This command is not supported in the prepared statement protocol yet'`

So I am wondering:

1. Am I correct that `$executeRawUnsafe` does send a prepared statement and not a text statement when within a transaction?
2. Is there any way to send it as a text statement?  Or is this approach not going to work?
3. Is there a better way to select a tenant database per query for mysql?  I have looked at this issue, but it appears that mysql work hasn't started (https://github.com/prisma/prisma/issues/1122).

Thanks for any help or thoughts.

### How to reproduce (optional)

From this client extension example, replace `forCompany()` with
https://github.com/prisma/prisma-client-extensions/blob/main/row-level-security/script.ts
```
function forCompany(companyId: string) {
  return Prisma.defineExtension((prisma) =>
    prisma.$extends({
      query: {
        $allModels: {
          async $allOperations({ args, query }) {
            const [, result] = await prisma.$transaction([
              prisma.$executeRawUnsafe('USE tenantA;'),
              query(args),
            ]);
            return result;
          },
        },
      },
    })
  );
}

```

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @mgraf-dozuki 👋 

Thank you for raising this question.

> 1. Am I correct that $executeRawUnsafe does send a prepared statement and not a text statement when within a transaction?

Yes, you are correct. The error message `Raw query failed. Code: '1295'. Message: 'This command is not supported in the prepared statement protocol yet'` indicates that `$executeRawUnsafe` is attempting to send the USE command as a prepared statement. This is consistent with the behavior described in this [GitHub issue](https://github.com/prisma/prisma/issues/3152), where certain commands like BEGIN, COMMIT, and ROLLBACK cannot be executed using the binary protocol and must use the text protocol instead.

> Is there any way to send it as a text statement? Or is this approach not going to work?

Currently, there is no direct way to force `$executeRawUnsafe` to send a query as a text statement within a transaction. There is an open [feature request](https://github.com/prisma/prisma/issues/3337) to support this in the future.


> Is there a better way to select a tenant database per query for MySQL?

You can take a look at the suggestion by another user in this open [issue](https://github.com/prisma/prisma/issues/12420#issuecomment-1363093740).

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
